### PR TITLE
Unify API errors processing

### PR DESCRIPTION
**Pull request by roman-mazur**
With this change, we define a custom error handler as express middleware
which renders 503 error with JSON response that includes status and message
fields.

The handler also logs the error, so the stack can be inspected in supervisor
logs. It's also a point where we can report the error to analytics services.

This removes a bunch of error handlers written in every request handler
function. Behaviour should remain unchanged except the fact that
`/healthy` endpoint now returns 503 in case of failure instead of 500.

Implemented while investigating problems with `GET /target-state` endpoint.

Change-type: patch
Signed-off-by: Roman Mazur <roman@balena.io>

### CODE REVIEW

File: src/device-api/v1.coffee [hunk 1]
```diff
@@ -47,16 +46,15 @@ exports.createV1Api = (router, applications) ->
 					return service
 			.then (service) ->
 				res.status(200).json({ containerId: service.containerId })
-		.catch (err) ->
-			res.status(503).send(err?.message or err or 'Unknown error')
+		.catch(next)
 
-	router.post '/v1/apps/:appId/stop', (req, res) ->
-		v1StopOrStart(req, res, 'stop')
+	router.post '/v1/apps/:appId/stop', (req, res, next) ->
```
Comment (Page-) [inline src/device-api/v1.coffee]: I think these would be better if you used a partial like
```suggestion
	router.post('/v1/apps/:appId/stop', createV1StopOrStart('start'))
```
or
```suggestion
	router.post('/v1/apps/:appId/stop', _.partial(v1StopOrStart, _, _, _, 'start'))
```

File: src/device-api/v2.ts [hunk 2]
```diff
@@ -85,286 +74,254 @@ export function createV2Api(router: Router, applications: ApplicationManager) {
...
+				.catch(next);
 		},
 	);
 
 	router.post(
 		'/v2/applications/:appId/restart-service',
-		(req: Request, res: Response) => {
-			return handleServiceAction(req, res, 'restart');
+		(req: Request, res: Response, next: NextFunction) => {
+			return handleServiceAction(req, res, next, 'restart');
```
Comment (Page-) [inline src/device-api/v2.ts]: Same for these, it'd be nice to use a partial to reduce duplication

Comment (roman-mazur) [reply src/device-api/v2.ts]: Thanks, good point.

Comment (roman-mazur) [reply src/device-api/v2.ts]: Addressed. PTAL again.

File: src/device-api/v1.coffee [hunk 3]
```diff
@@ -47,16 +46,14 @@ exports.createV1Api = (router, applications) ->
...
-			res.status(503).send(err?.message or err or 'Unknown error')
+		.catch(next)
 
-	router.post '/v1/apps/:appId/stop', (req, res) ->
-		v1StopOrStart(req, res, 'stop')
+	createV1StopOrStartHandler = (action) -> _.partial(v1StopOrStart, _, _, _, action)
 
-	router.post '/v1/apps/:appId/start', (req, res) ->
-		v1StopOrStart(req, res, 'start')
+	router.post '/v1/apps/:appId/stop', createV1StopOrStartHandler('stop')
```
Comment (CameronDiver) [inline src/device-api/v1.coffee]: Please add parens for single line functions calls in coffeescript

Comment (roman-mazur) [reply src/device-api/v1.coffee]: Done. But why doesn't our linter require this?

**Comment by roman-mazur** [comment]
@balena-ci retest, please

**Comment by roman-mazur** [comment]
aarch64-debug build hangs on making sqlite... restarted

**Comment by CameronDiver** [comment]
@roman-mazur it's not hanging, it just takes a bunch of time due to qemu.

**Comment by roman-mazur** [comment]
aarch64-debug took 47:36 - my personal record for CI builds :)

File: src/device-api/v1.coffee (hunk 3 above)
Comment (Page-) [reply src/device-api/v1.coffee]: You can either enforce parens in all cases or none @roman-mazur, what we decided on for coffeescript was the more nuanced "always for functions without a callback/single line callback, optional/never for functions with a multiline callback" - overall that makes the best use of significant whitespace in coffeescript whilst avoid a lot of the confusion around things like `a b c, d, e, f (g h), (i j)` where things can get crazy with various interactions of parens and whitespace and it's hard to parse without really thinking about it (that example compiles to `a(b(c, d, e, f(g(h), i(j))))`)
Either way we're steadily moving to typescript which removes the coffeescript linting capabilities as an issue :)
